File: javascript/sliding-window/longest_substring_with_k_distinct.js
const longest_substring_with_k_distinct = function(str, k) {
  let windowStart = 0, maxLength = -Infinity, freq = {};

  for (let windowEnd = 0; windowEnd < str.length; windowEnd++) {
    let rightChar = str[windowEnd];

    if (!(rightChar in freq)) freq[rightChar] = 0;
    freq[rightChar]++;

    // If we are over K then let's shrink the window
    while (Object.keys(freq).length > k) {
      // Want to make sure we have <= K keys
      let leftChar = str[windowStart];
      freq[leftChar]--;
      if (freq[leftChar] === 0) delete freq[leftChar];

      // Move up the start of the window
      windowStart++;
    }

    // Calculate the maxLength now that we've met out criteria of <= K keys
    maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
  }
  
  return maxLength;
};

File: python/sliding-window/longest_substring_with_k_distinct.py
def longest_substring_with_k_distinct(str1, k):
  start, max_length, freq = 0, float('-inf'), {}

  for end in range(len(str1)):
    rightElem = str1[end]

    if rightElem not in freq:
      freq[rightElem] = 0
    freq[rightElem] += 1

    while (len(freq) > k):
      leftElem = str1[start]
      freq[leftElem] -= 1

      if freq[leftElem] == 0:
        del freq[leftElem]

      start += 1

    max_length = max(max_length, end - start + 1)

  return max_length